hubble/recorder: Refactor service implementation to fix multiple races

gandro opened the pull request:
This PR fixes multiple concurrency issues in the implementation of the
Hubble Recorder API. The main one being that we were sending responses
to the client from both the main `Record` function, as well as the
`watchRecording` function which was spawned in a separate go routine.

However, sending to a grpc.ServerStream from multiple go routines is
_not_ safe: https://pkg.go.dev/google.golang.org/grpc#ServerStream
It is however safe to have one go routine receive from, and another go routine
send to the stream.

Therefore, this commit restructures the Hubble Recorder API in such a
way that only the `Record` stub ever sends back messages to the client.
Receiving is done in a separate go routine which forwards all received
messages into a channel, allowing us to select on incoming responses.

In addition, this commit hopefully also makes the logic a bit more
easier to read, as it tries to separate the cleanup of resources and
communicating with the client a bit more explicitly. It also drastically
simplifies the implementation of a follow-up PR which extends the
API with stop conditions (#16473).

I apologize for the size, but I had to restructure quite a bit to fix
the underlying issue. **I highly recommended to review per commit.**

gandro committed:
hubble/recorder: Allow multiple go routines to wait for sink

This commit splits the sink's `done chan error` channel (which only
allows a single consumer to wait on the sink to finish) into a `chan
struct{}` channel and a `lastError error` variable.

This enables us to signal that the sink has finished by closing the
channel instead of sending a value over it. Closing the channel allows
multiple go routines to block on this event via `<-s.done`. The final
error value can then be retrieved via `s.err()`. This pattern is very
similar to how `context.Context` works.

This commit does not yet make use of this functionality. The changes
enabled by this will follow in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/recoder: Automatically unregister sink when stopped

This commit changes the interface of sink.Dispatch from an explicit
`RegisterSink`+`UnregisterSink` pair to a `StartSink` call which will
unregister itself when it stops due to an error, an expired context, or
an explicit stop request.

This commit does not introduce any functional changes, it is purely a
refactoring.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/recorder: Improve ergonomics by not closing stats channel

Since we have introduced the `Handle.Done` channel, we do not have toV
signal the shutdown of the sink by closing the statistics channel
anymore. Instead, consumers can now wait on the `Handle.Done` channel
getting closed.

While there are not many benefits in this version of the code, it will
make the select statement in a subsequent commit much more readable.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/recorder: Fix grpc send from concurrent go routine

This commit fixes a concurrency issue in the implementation of the
Hubble Recorder API. Before this commit, we were sending responses to
the client from both the main `Record` function, as well as the
`watchRecording` function which was spawned in a separate go routine.

However, sending to a grpc.ServerStream from multiple go routines is
_not_ safe: https://pkg.go.dev/google.golang.org/grpc#ServerStream It is
however safe to have one go routine receive from, and another go routine
send to the stream.

Therefore, this commit restructures the Hubble Recorder API in such a
way that only the `Record` stub ever sends back messages to the client.
Receiving is done in a separate go routine which forwards all received
messages into a channel, allowing us to select on incoming responses.

In addition, this commit hopefully also makes the logic a bit more
easier to read, as it tries to separate the cleanup of resources and
communicating with the client a bit more explicitly.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/recorder: Stop without draining queue

Previously, we waited for the recording queue to drain when the client
requested a stop. However, because the client has no visibility into the
queue (and indeed doesn't even know if there are queued records when
they issue a stop request), this does not provide any value to the
client.

Therefore, this PR changes the semantics of a stop request by
immediately initiating a shutdown, instead of waiting for the queue to
drain. This ensures that the resulting recording more closely matches
the observed statistics at the time when the client issued a stop
request. It also simplifies the code a bit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

michi-covalent reviewed (approved):
:shipit:

tklauser commented on pkg/hubble/recorder/sink/sink.go:
```diff
@@ -25,67 +24,72 @@ import (
 	go func() {
-		// this defer executes w.Close(), but also makes sure err
-		// is sent to the done channel upon exit
+		// this defer executes w.Close(), but also makes sure set lastError and
```
Nit: 
```suggestion
		// this defer executes p.Write.Close(), but also makes sure to set lastError and
```

tklauser commented on pkg/hubble/recorder/sink/sink.go:
```diff
@@ -25,67 +24,72 @@ import (
+			} else {
+				s.lastError = err
+			}
+			close(s.done)
```
It looks like all receivers (via `Handle.Done`) don't take `s.mutex`. I guess technically we could move the `close(s.done)` out of the critical section. Or could this lead to problems?

gandro replied:
Good catch - yes, we can move it out of the critical section! I think I'll also add a comment to document what `s.mutex` actually protects at this point.

tklauser reviewed (approved):
Well structured PR was nice to review and the inline comments help a great deal understanding the control flow. Thanks!

Two small nits inline.

gandro committed:
hubble/recorder: Be more explicit about mutex

This commit documents what fields are now protected by the mutex in
`type sink` and updates a two usages accordingly, by moving channel
operations out of the critical section.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser reviewed (approved):
Thanks :rocket:
